landscaper/gardenlet/pkg/gardenlet_reconcile.go

```diff
@@ -0,0 +1,266 @@
+			seedKubeconfig = []byte(g.gardenletConfiguration.SeedClientConnection.Kubeconfig)
+		}
+
+		if err := g.deploySeedSecret(ctx, seedKubeconfig, g.gardenletConfiguration.SeedConfig.Spec.SecretRef); err != nil {
+			return fmt.Errorf("failed to deploy secret for the Seed resource containing the Seed cluster's kubeconfig: %v", err)
+		}
+	}
+
+	// if configured, deploy the seed-backup secret to the Garden cluster
+	if g.Imports.SeedBackup != nil && seedConfig.Spec.Backup != nil {
+		credentials := make(map[string][]byte)
+		marshalJSON, err := g.Imports.SeedBackup.Credentials.MarshalJSON()
+		if err != nil {
+			return err
+		}
+
+		if err := json.Unmarshal(marshalJSON, &credentials); err != nil {
+			return err
+		}
+
+		if err := g.deployBackupSecret(ctx, g.Imports.SeedBackup.Provider, credentials, seedConfig.Spec.Backup.SecretRef); err != nil {
+			return fmt.Errorf("failed to deploy the Seed backup secret to the Garden cluster: %v", err)
+		}
+	}
+
+	// This check is to determine whether the Gardenlet needs bootstrap credentials.
+	// It is not enough to just check for the existence of the Seed resource to verify that the current Gardenlet
+	// has already valid Garden cluster credentials.
+	// The Gardenlet might have failed to rotate its credentials and fails to reconcile the Seed.
+	// Possibly, this deploys bootstrap credentials even though they are not required in order to increase reliability
+	// this should not be a problem, as the landscaper always uses bootstrap secrets with a limited lifetime.
+	isAlreadyBootstrapped, err := g.isSeedBootstrapped(ctx, seedConfig.ObjectMeta)
+	if err != nil {
+		return fmt.Errorf("failed to check if seed %q is already bootstrapped: %v", seedConfig.ObjectMeta.Name, err)
+	}
+
+	var bootstrapKubeconfig []byte
+	if !isAlreadyBootstrapped {
+		bootstrapKubeconfig, err = g.getKubeconfigWithBootstrapToken(ctx, seedConfig.ObjectMeta.Name)
+		if err != nil {
+			return fmt.Errorf("failed to compute the bootstrap kubeconfig: %v", err)
+		}
+	}
+
+	gardenNamespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.GardenNamespace}}
+	if err := g.seedClient.Client().Create(ctx, gardenNamespace); err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	values := g.computeGardenletChartValues(bootstrapKubeconfig)
+
+	applier := applier.NewGardenletChartApplier(g.seedClient.ChartApplier(), values, GetChartPath())
+	if err := applier.Deploy(ctx); err != nil {
+		return fmt.Errorf("failed deploying Gardenlet chart to the Seed cluster: %v", err)
+	}
+
+	g.log.Infof("Successfully deployed Gardenlet resources for Seed %q", g.gardenletConfiguration.SeedConfig.Name)
+
+	return g.waitForRolloutToBeComplete(ctx)
+}
+
+// Sleep returns the time the process should sleep to give the gardenlet process
+// time to startup and either fail or proceed
+// exposed for testing
+var Sleep = func() time.Duration { return 10 * time.Second }
```

Comment:
Same question here, could you perhaps use a `Landscaper` field instead of exposing this function?

Reply:
In this case it is something use only during the deletion so I would not add it as a field to the landscaper struct.

Reply:
Why the fact that "it is something use only during the deletion" should even matter? It's still a value that you would like to set differently for different use cases (e.g. production vs. testing), e.g. `deletionSleepDuration`.

Reply:
mhm I cannot follow, sorry. 
To me this comes down to style and I would either create a dedicated DeletionLandscaper struct that includes a field _deletionSleepDuration_ or define it like it is today as a separate variable outside.
It seems odd to me to have a field in the struct that is simply not used.

But if you have a strong opinion here, I can change it.

Reply:
It seems like a misunderstanding, let me clarify. What I meant is simply:
```
time.Sleep(g.rolloutSleepDuration)
```
This field is used during rollout, not deletion, so I renamed it accordingly. And of course it's used here, exactly once. For testing, you could just avoid setting it since 0 is what you want. For production, you should set it to an appropriate (perhaps configurable?) value via the constructor. I think this approach is much cleaner than having global function variables to be overriden by tests, and should be used consistently whenever possible.